modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/TestRunner.java

```diff
@@ -93,197 +93,286 @@ public void runTest( ITestRunnerDataProvider dataProvider  ) throws TestRunExcep
+            logger.error(msg+" "+e.getMessage());
+            reportEnvFailFinishedResult(e);
+            throw new TestRunException(msg,e);
+        }
+        return testClassWrapper;
+    }
 
-                logger.debug("Stopping heartbeat...");
+    private boolean isAnyReasonToIgnoreTests(ITestRunManagers managers) throws TestRunException {
+        boolean isIgnore = false ;
+        try {
+            if (managers.anyReasonTestClassShouldBeIgnored()) {
+                logger.debug("managers.anyReasonTestClassShouldBeIgnored() is true. Shutting down.");
                 stopHeartbeat();
+                updateStatus(TestRunLifecycleStatus.FINISHED, "finished");
+                isIgnore = true; // TODO handle ignored classes
+            }
+        } catch (Exception e) {
+            String msg = "Problem asking Managers for an ignore reason";
+            logger.error(msg);
+            reportEnvFailFinishedResult(e);
+            throw new TestRunException(msg, e);
+        }
+        return isIgnore;
+    }
 
-                // Record all the CPS properties that were accessed
-                saveUsedCPSPropertiesToArtifact(this.framework.getRecordProperties(), this.fileSystem, this.ras);
-                // And all the overrides the test was passed.
-                saveAllOverridesPassedToArtifact(overrideProperties, this.fileSystem , this.ras);
-
-                // *** If this was a local run, then we will want to remove the run properties
-                // from the DSS immediately
-                // *** for automation, we will let the core manager clean up after a while
-                // *** Local runs will have access to the run details via a view,
-                // *** But automation runs will only exist in the RAS if we delete them, so need
-                // to give
-                // *** time for things like jenkins and other run requesters to obtain the
-                // result and RAS id before
-                // *** deleting, default is to keep the automation run properties for 5 minutes
-                if (!markedWaiting) {
-                    deleteRunProperties(this.framework);
-                }
-            } else if (this.runType == RunType.SHARED_ENVIRONMENT_BUILD) {
+    private ITestRunManagers initialiseManagers(Class<?> testClass , ITestRunnerDataProvider dataProvider) throws TestRunException {
+        // *** Initialise the Managers ready for the test run
+        ITestRunManagers managers ;
+        try {
+            GalasaTest galasaTest = new GalasaTest(testClass);
+            managers = dataProvider.createTestRunManagers(galasaTest);
+        } catch (Exception e) {
+            // Managers are custom code, may be prone to failure if they are immature...
+            // so catch any exception and turn it into a TestRunException.
+            String msg = "Exception Exception caught. "+e.getMessage()+" Shutting down and Re-throwing.";
```

Comment:
Duplicate mention of "Exception" in this msg

Reply:
fixed